watertap/property_models/unit_specific/activated_sludge/asm3_reactions.py

```diff
@@ -0,0 +1,863 @@
+        )
+        self.K_A_ALK = pyo.Var(
+            initialize=0.5e-3,
+            units=pyo.units.kmol / pyo.units.m**3,
+            domain=pyo.PositiveReals,
+            doc="Bicarbonate saturation for nitrifiers (kmol-HCO3- / m3)",
+        )
+        b_A_O2_dict = {"10C": 0.05, "20C": 0.15}
+        self.b_A_O2 = pyo.Var(
+            b_A_O2_dict.keys(),
+            domain=pyo.PositiveReals,
+            initialize=b_A_O2_dict,
+            units=pyo.units.day**-1,
+            doc="Aerobic endogenous respiration rate of X_A (day^-1)",
+        )
+        b_A_NOX_dict = {"10C": 0.02, "20C": 0.05}
+        self.b_A_NOX = pyo.Var(
+            b_A_NOX_dict.keys(),
+            domain=pyo.PositiveReals,
+            initialize=b_A_NOX_dict,
+            units=pyo.units.day**-1,
+            doc="Anoxic endogenous respiration rate of X_A (day^-1)",
+        )
+
+        # Stoichiometric numbers from Table 1
+        # obtained by \sum_i^12 νji*ikI
+        x1 = 1.0 - self.f_SI
+        x2 = -1.0 + self.Y_STO_O2
+        x3 = (-1.0 + self.Y_STO_NOX) / (64.0 / 14.0 - 24.0 / 14.0)
+        x4 = 1.0 - 1.0 / self.Y_H_O2
+        x5 = (+1.0 - 1.0 / self.Y_H_NOX) / (64.0 / 14.0 - 24.0 / 14.0)
+        x6 = -1.0 + self.f_XI
+        x7 = (self.f_XI - 1.0) / (64.0 / 14.0 - 24.0 / 14.0)
+        x8 = -1.0
+        x9 = -1.0 / (64.0 / 14.0 - 24.0 / 14.0)
+        x10 = -(64.0 / 14.0) / self.Y_A + 1.0
+        x11 = self.f_XI - 1.0
+        x12 = (self.f_XI - 1.0) / (64.0 / 14.0 - 24.0 / 14.0)
+
+        y1 = -self.f_SI * self.i_NSI - (1.0 - self.f_SI) * self.i_NSS + self.i_NXS
+        y2 = self.i_NSS
+        y3 = self.i_NSS
+        y4 = -self.i_NBM
+        y5 = -self.i_NBM
+        y6 = -self.f_XI * self.i_NXI + self.i_NBM
+        y7 = -self.f_XI * self.i_NXI + self.i_NBM
+        y10 = -1.0 / self.Y_A - self.i_NBM
+        y11 = -self.f_XI * self.i_NXI + self.i_NBM
+        y12 = -self.f_XI * self.i_NXI + self.i_NBM
+
+        z1 = y1 / 14.0
+        z2 = y2 / 14.0
+        z3 = y3 / 14.0 - x3 / 14.0
+        z4 = y4 / 14.0
+        z5 = y5 / 14.0 - x5 / 14.0
+        z6 = y6 / 14.0
+        z7 = y7 / 14.0 - x7 / 14.0
+        z9 = -x9 / 14.0
+        z10 = y10 / 14.0 - 1.0 / (self.Y_A * 14.0)
+        z11 = y11 / 14.0
+        z12 = y12 / 14.0 - x12 / 14.0
+
+        t1 = -self.i_SSXS
+        t2 = self.Y_STO_O2 * self.i_SSSTO
+        t3 = self.Y_STO_NOX * self.i_SSSTO
+        t4 = self.i_SSBM - 1.0 / self.Y_H_O2 * self.i_SSSTO
+        t5 = self.i_SSBM - 1.0 / self.Y_H_NOX * self.i_SSSTO
+        t6 = self.f_XI * self.i_SSXI - self.i_SSBM
+        t7 = self.f_XI * self.i_SSXI - self.i_SSBM
+        t8 = -self.i_SSSTO
+        t9 = -self.i_SSSTO
+        t10 = self.i_SSBM
+        t11 = self.f_XI * self.i_SSXI - self.i_SSBM
+        t12 = self.f_XI * self.i_SSXI - self.i_SSBM
+
+        # Reaction Stoichiometry
+        # This is the stoichiometric part the Peterson matrix in dict form
```

Comment:
Is this matrix on Box somewhere?

Reply:
Yeah, the original paper is here: [https://app.box.com/s/a0b5ghfttqr0qh6916f1bs2w9qy8iox6](url)
And there is a review paper: [https://app.box.com/s/0ig1txxlni6jyilbe4bwdu4glc4dxttu](url)